working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+
+See also:
+
+* [JEP 453 - Structured Concurrency](https://openjdk.org/jeps/453)
+* [Wikipedia - Structured Concurrency](https://en.wikipedia.org/wiki/Structured_concurrency)
+
+
+
+### `dart:concurrent`
+
+`dart:concurrent` will serve as a library hosting low-level concurrency primitives. 
+
+#### `Thread` and `ThreadPool`
+
+Isolates are *not* threads even though they are often confused with ones. A code running within an isolate might be executing on a dedicated OS thread or it might running on a dedicated pool of threads. When expanding Dart's multithreading capabilities it seems reasonable to introduce more explicit ways to control threads. 
+
+```dart
+// dart:concurrent
+
+abstract class Thread implements Shareable {
+  /// Runs the given function in a new thread.
+  ///
+  /// The function is run in a shared isolate, meaning that 
+  /// it will not have access to the non-`@shared` state.
+  ///
+  /// The function will be run in a `Zone` which uses the
+  /// spawned thread as an executor for all callbacks: this
+  /// means the thread will remain alive as long as there is 
+  /// a callback referencing it.
+  external static Thread start<T>(FutureOr<T> Function() main); 
+  
+  external Thread get current;
+    
+  Future<void> join();
+  
+  void interrupt();
+    
+  external set priority(ThreadPriority value);
+  ThreadPriority get priority;
+}
+
+abstract class ThreadPool implements Shareable {
+  external factory ThreadPool({required int concurrency});
+  
+  /// Run the given function on this thread pool.
+  /// 
+  /// The [task] is executed in a shared isolate of the current
+  /// isolate group. 
+  ///
+  /// The function will be run in a `Zone` which uses this
+  /// pool as an executor for all callbacks: this
+  /// means the pool will remain alive as long as there is 
+  /// a callback referencing it.
+  void postTask(void Function() task);
+}
+```
+
+We might also want to provide a way to synchronously execute code in a specific isolate on the _current_ thread:
+
+```dart
+class Isolate {
+    T runSync<T>(T Function() cb);
+}
+```
+
+##### Asynchronous code and threads
+
+Consider the following example:
+
+``` dart
+Thread.start(() async {
+  var v = await foo();
+  var u = await bar();
+});
+```
+
+Connecting execution / scheduling behavior to `Zone` allows us to give a clear semantics to this code: this code will run on a specific (newly spawned) thread and will not change threads between suspending and resumptions.  
+
+#### Atomic operations
+
+`Atomic<T>` is a wrapper around a value of type `T` which can be updated atomically. It can only be used with true reference types - an attempt to create an `AtomicRef<int>`, `AtomicRef<double>` , `AtomicRef<(T1, ..., Tn)>` will throw.  
```

Comment:
Should this be `AtomicRef<T>`?

```suggestion
`AtomicRef<T>` is a wrapper around a value of type `T` which can be updated atomically. It can only be used with true reference types - an attempt to create an `AtomicRef<int>`, `AtomicRef<double>` , `AtomicRef<(T1, ..., Tn)>` will throw.  
```